--- makepkg.conf	2018-05-30 08:38:22.167804726 +0200
+++ makepkg.conf	2018-05-30 08:38:02.885358183 +0200
@@ -113,15 +113,15 @@
 #-- Destination: specify a fixed directory where all packages will be placed
 #PKGDEST=/home/packages
 #-- Source cache: specify a fixed directory where source files will be cached
-#SRCDEST=/home/sources
+SRCDEST='/home/makepkg/packageSources'
 #-- Source packages: specify a fixed directory where all src packages will be placed
 #SRCPKGDEST=/home/srcpackages
 #-- Log files: specify a fixed directory where all log files will be placed
 #LOGDEST=/home/makepkglogs
 #-- Packager: name/email of the person or organization building packages
-PACKAGER="Erich Eckner <arch32 at eckner dot net>"
+PACKAGER='Erich Eckner <arch at eckner dot net>'
 #-- Specify a key to use for package signing
-#GPGKEY=""
+GPGKEY='5FDCA472AB93292BC678FD59255A76DB9A12601A'
 
 #########################################################################
 # COMPRESSION DEFAULTS
