--- makepkg-i686.conf	2018-05-30 08:38:02.885358183 +0200
+++ makepkg-i686.conf	2018-05-30 08:40:25.575856939 +0200
@@ -32,13 +32,13 @@
 # ARCHITECTURE, COMPILE FLAGS
 #########################################################################
 #
-CARCH="x86_64"
-CHOST="x86_64-pc-linux-gnu"
+CARCH="i686"
+CHOST="i686-pc-linux-gnu"
 
 #-- Compiler and Linker Flags
 CPPFLAGS="-D_FORTIFY_SOURCE=2"
-CFLAGS="-march=x86-64 -mtune=generic -O2 -pipe -fno-plt"
-CXXFLAGS="-march=x86-64 -mtune=generic -O2 -pipe -fno-plt"
+CFLAGS="-m32 -march=i686 -mtune=generic -O2 -pipe -fno-plt"
+CXXFLAGS="-m32 -march=i686 -mtune=generic -O2 -pipe -fno-plt"
 LDFLAGS="-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now"
 #-- Make Flags: change this for DistCC/SMP systems
 MAKEFLAGS="-j4"
