--- makepkg.conf	2016-04-07 00:16:07.721500790 +0200
+++ makepkg.conf	2016-04-13 09:43:35.042178024 +0200
@@ -113,15 +113,15 @@
 #-- Destination: specify a fixed directory where all packages will be placed
 #PKGDEST=/home/packages
 #-- Source cache: specify a fixed directory where source files will be cached
-#SRCDEST=/home/sources
+SRCDEST=/home/makepkg/packageSources
 #-- Source packages: specify a fixed directory where all src packages will be placed
 #SRCPKGDEST=/home/srcpackages
 #-- Log files: specify a fixed directory where all log files will be placed
 #LOGDEST=/home/makepkglogs
 #-- Packager: name/email of the person or organization building packages
-PACKAGER="Erich Eckner <arch32 at eckner dot net>"
+PACKAGER="Erich Eckner <arch@eckner.net>"
 #-- Specify a key to use for package signing
-#GPGKEY=""
+GPGKEY="7C98C4C3DE926168DC46FBAA3D06644243BF68D3"
 
 #########################################################################
 # COMPRESSION DEFAULTS
