--- a/trunk/findversion.sh	2016-04-04 13:46:56.085550989 +0200
+++ b/trunk/findversion.sh	2016-04-04 14:01:08.537965131 +0200
@@ -128,10 +128,12 @@
 	exit
 else
 	# We don't know
-	MODIFIED="1"
-	BRANCH=""
-	REV=""
-	REV_NR=""
+	MODIFIED="2"
+        HASH="$(find $ROOT_DIR/{src,bin} -type f -exec sha512sum {} \; | sha512sum - | cut)"
+	BRANCH="r27534-hard-game-by-ewe"
+	REV="g`echo $HASH | cut -c1-8`"
+	REV_NR="27534"
+	TAG=""
 fi
 
 if [ "$MODIFIED" -eq "2" ]; then
--- a/trunk/config.lib	2016-04-04 14:04:14.881840501 +0200
+++ b/trunk/config.lib	2016-04-04 14:04:25.561871556 +0200
@@ -1040,32 +1040,6 @@
 		fi
 	fi
 
-	if [ -d "$ROOT_DIR/.svn" ] && [ -n "`svn help 2>/dev/null`" ]; then
-		log 1 "checking revision... svn detection"
-	elif [ -d "$ROOT_DIR/../.svn" ] && [ -n "`svn help 2>/dev/null`" ] && [ -n "`LC_ALL=C svn info $ROOT_DIR/.. | grep '^URL:.*tags$'`" ]; then
-		# subversion changed its behaviour; now not all folders have a .svn folder,
-		# but only the root folder. Since making tags requires a (sparse) checkout
-		# of the tags folder, the folder of the tag does not have a .svn folder
-		# anymore and this fails to detect the subversion repository checkout.
-		log 1 "checking revision... svn detection (tag)"
-	elif [ -d "$ROOT_DIR/.git" ] && [ -n "`git help 2>/dev/null`" ]; then
-		log 1 "checking revision... git detection"
-	elif [ -d "$ROOT_DIR/.hg" ] && [ -n "`HGPLAIN= hg help 2>/dev/null`" ]; then
-		log 1 "checking revision... hg detection"
-	elif [ -f "$ROOT_DIR/.ottdrev" ]; then
-		log 1 "checking revision... source tarball"
-	else
-		log 1 "checking revision... no detection"
-		log 1 "WARNING: there is no means to determine the version."
-		log 1 "WARNING: please use a subversion, mercurial, or git checkout of OpenTTD."
-		log 1 "WARNING: you can only join game servers that have been compiled without"
-		log 1 "WARNING:   version detection."
-		log 1 "WARNING: there is a great chance you desync."
-		log 1 "WARNING: USE WITH CAUTION!"
-
-		sleep 5
-	fi
-
 	if [ "$doc_dir" = "1" ]; then
 		if [ "$os" = "UNIX" ] || [ "$os" = "FREEBSD" ] || [ "$os" = "DRAGONFLY" ] || [ "$os" = "OPENBSD" ] || [ "$os" = "NETBSD" ] || [ "$os" = "HPUX" ] || [ "$os" = "SUNOS" ]; then
 			doc_dir="share/doc/openttd"
