--- fetchmail-6.4.1/socket.c	2019-09-15 17:48:53.000000000 +0200
+++ fetchmail-6.4.1/socket.c	2019-09-30 07:13:58.260770910 +0200
@@ -905,13 +905,8 @@
 	}
 
 	if (!strcasecmp("ssl3", *myproto)) {
-#if (HAVE_DECL_SSLV3_CLIENT_METHOD > 0) && (0 == OPENSSL_NO_SSL3 + 0)
-		_ctx[sock] = SSL_CTX_new(SSLv3_client_method());
-		*avoid_ssl_versions &= ~SSL_OP_NO_SSLv3;
-#else
 		report(stderr, GT_("Your OpenSSL version does not support SSLv3.\n"));
 		return -1;
-#endif
 	} else if (!strcasecmp("ssl3+", *myproto)) {
 		*avoid_ssl_versions &= ~SSL_OP_NO_SSLv3;
 		*myproto = NULL;
