--- a/setup.py	2018-07-10 16:33:03.044849216 +0200
+++ b/setup.py	2018-07-10 16:32:51.198144418 +0200
@@ -197,7 +197,7 @@
     requirements = f.readlines()
 
 # now we have all the data files, so we can run setup
-setup(name = 'xraylarch',
+dist = setup(name = 'xraylarch',
       version = __version__,
       author = 'Matthew Newville and the X-rayLarch Development Team',
       author_email = 'newville@cars.uchicago.edu',
@@ -215,6 +215,7 @@
                    'Programming Language :: Python',
                    'Topic :: Scientific/Engineering'],
      )
+icmd = dist.get_command_obj('install')
 
 
 def remove_cruft(basedir, filelist):
@@ -307,7 +308,7 @@
     scriptdir = pjoin(sys.exec_prefix, bindir)
     for src in scripts:
         _, fname = psplit(src)
-        dest = pjoin(scriptdir, fname)
+        dest = pjoin(icmd.root + scriptdir, fname)
         shutil.copy(src, dest)
         os.chmod(dest, 493) # mode=755
 
