--- a/examples/nginx/acme2certifier.ini	2020-05-24 13:46:24.045480702 +0200
+++ b/examples/nginx/acme2certifier.ini	2020-05-24 12:52:35.875846711 +0200
@@ -1,10 +1,13 @@
 [uwsgi]
+chdir = /var/lib/acme2certifier
 module = acme2certifier_wsgi:application
+plugins = python
+wsgi-file = /var/lib/acme2certifier/acme2certifier_wsgi.py
 master = true
 processes = 5
-uid = nginx
+uid = http
 socket = /run/uwsgi/acme.sock
-chown-socket = nginx
+chown-socket = http
 chmod-socket = 660
 vacuum = true
 die-on-term = true
--- a/examples/nginx/uwsgi.service	2020-05-24 16:26:36.289592292 +0200
+++ b/examples/nginx/uwsgi.service	2020-05-24 13:44:56.156306673 +0200
@@ -3,11 +3,11 @@
 
 [Service]
 RuntimeDirectory=uwsgi
-ExecStart=/usr/bin/bash -c 'cd /opt/acme2certifier; uwsgi --ini acme2certifier.ini'
+ExecStart=/usr/bin/uwsgi --ini /etc/uwsgi/acme2certifier.ini
 Restart=always
 Type=notify
 NotifyAccess=all
-User=nginx
+User=http
 
 [Install]
 WantedBy=multi-user.target
