diff --git a/src/graph.c b/src/graph.c
index c63fdf9c..ecabb1bf 100644
--- a/src/graph.c
+++ b/src/graph.c
@@ -366,7 +366,12 @@ void dump_graph(void) {
 	/* now dump all edges */
 	for(node = edge_weight_tree->head; node; node = node->next) {
 		e = node->data;
-		fprintf(file, "	\"%s\" -> \"%s\";\n", e->from->name, e->to->name);
+		if (e->reverse) {
+			if (e->reverse > e)
+				fprintf(file, "	\"%s\" -> \"%s\" [dir=both];\n", e->from->name, e->to->name);
+		}
+		else
+			fprintf(file, "	\"%s\" -> \"%s\";\n", e->from->name, e->to->name);
 	}
 
 	fprintf(file, "}\n");
