--- wslay-release-1.1.1/doc/sphinx/conf.py.in	2021-05-17 20:46:23.440993958 +0200
+++ wslay-release-1.1.1/doc/sphinx/conf.py.in	2021-05-17 20:45:05.563332347 +0200
@@ -123,6 +123,8 @@
 # so a file named "default.css" will overwrite the builtin "default.css".
 html_static_path = ['_static/default2.css']
 
+html_css_files = ['default2.css']
+
 # If not '', a 'Last updated on:' timestamp is inserted at every page bottom,
 # using the given strftime format.
 #html_last_updated_fmt = '%b %d, %Y'
@@ -302,4 +304,4 @@
     ]
 
 def setup(app):
-    app.add_stylesheet('default2.css')
+    pass
