--- a/etc/pam.d/system-auth	2021-06-17 18:57:25.959996395 +0200
+++ b/etc/pam.d/system-auth	2021-06-17 18:58:31.533329725 +0200
@@ -11,12 +11,10 @@
 # If you drop the above call to pam_faillock.so the lock will be done also
 # on non-consecutive authentication failures.
 
--account   [success=1 default=ignore]  pam_systemd_home.so
 account    required                    pam_unix.so
 account    optional                    pam_permit.so
 account    required                    pam_time.so
 
--password  [success=1 default=ignore]  pam_systemd_home.so
 password   required                    pam_unix.so          try_first_pass nullok shadow sha512
 password   optional                    pam_permit.so
 
