--- conky-1.11.0/src/top.h	2018-12-02 19:26:02.000000000 +0100
+++ conky-1.11.0/src/top.h	2018-12-02 22:11:18.515668790 +0100
@@ -73,7 +73,7 @@
  * and it'll take me a while to write a replacement. */
 #define BUFFER_LEN 1024
 
-#define MAX_SP 10  // number of elements to sort
+#define MAX_SP 20  // number of elements to sort
 
 /******************************************
  * Process class						  *
