diff -ru3 conky-1.10.1/src/top.h ../conky-1.10.1/src/top.h
--- conky-1.10.1/src/top.h	2015-11-18 16:47:45.000000000 +0100
+++ ../conky-1.10.1/src/top.h	2015-11-20 20:29:25.993328891 +0100
@@ -75,7 +75,7 @@
  * and it'll take me a while to write a replacement. */
 #define BUFFER_LEN 1024
 
-#define MAX_SP 10	// number of elements to sort
+#define MAX_SP 20	// number of elements to sort
 
 /******************************************
  * Process class						  *
